### PR TITLE
74 toolbar

### LINKED ISSUE #74: Toolbar
There is a menu for selecting which structure to build.

### PR DESCRIPTION
**Pull request by Matte02**
## Summary
Add a Toolbar, and structures button on the toolbar
image

## Test Coverage
Have not changed since I've only changed Views, Controllers and the application.

## Diagrams
Will add these later.

Closes #74

### CODE REVIEW

File: src/main/java/com/thebois/Pawntastic.java [hunk 1]
```diff
@@ -36,8 +37,9 @@
     private static final float VIEWPORT_HEIGHT = 1000;
     private static final int DEFAULT_FONT_SIZE = 26;
     private static final int PAWN_POSITIONS = 50;
-    private static final int TILE_SIZE = (int) Math.min(VIEWPORT_HEIGHT, VIEWPORT_WIDTH)
-                                         / WORLD_SIZE;
+    private static final int TOOLBAR_HEIGHT = 100;
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/Pawntastic.java:40]: Only noticed this here first, but should `Pawntastic` know the toolbar (and other rendered elements) dimensions?

File: src/main/java/com/thebois/Pawntastic.java [hunk 2]
```diff
@@ -87,6 +90,15 @@ public static EventBus getEventBus() {
...
     }
 
+    /**
+     * Gets the height of the tool bar.
+     *
+     * @return The height.
+     */
+    public static int getToolBarHeight() {
+        return TOOLBAR_HEIGHT;
+    }
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/Pawntastic.java:100]: My first comment is related to this, feels a bit wrong to have it here, no?

**Review by jonathan-lindqvist — CHANGES_REQUESTED** [review]
Would lie if I said I understand exactly how this all works, but I think I got the overall concept, nicely done! 

Left two comments in pawntastic that you need to look over.

File: src/main/java/com/thebois/Pawntastic.java (hunk 1 above)
Comment (Matte02) [reply src/main/java/com/thebois/Pawntastic.java:40]: Don't really know where to move though, since GameScreen, some views need the Toolbarheight, and most views and some controllers need the Tile size. 

I guess it's all about what do want everything to depend on, Application or some other class where we would move these.

Comment (martinjonsson01) [reply src/main/java/com/thebois/Pawntastic.java:40]: Shouldn't only `ToolbarView` need this height? `GameScreen` should just be able to get the height of the actor, no?

File: src/main/java/com/thebois/controllers/game/StructureController.java [hunk 3]
```diff
@@ -1,76 +1,38 @@
...
      * Creates a instance of a Structure Controller.
      *
-     * @param world     The world in which the structures exists.
-     * @param projector Projector used for converting screen coordinates to world coordinates.
+     * @param world The world in which the structures exists.
      */
     public StructureController(
-        final World world, final IProjector projector) {
-
+        final World world) {
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/controllers/game/StructureController.java:22]: style

Now that there are so few parameters, you can remove the line wrapping here.

File: src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        structureToolbarView = new StructureToolbarView(skin, this);
+    }
+
+    @Override
+    public boolean touchDown(final int x, final int y, final int pointer, final int button) {
+        if (button == LEFT_CLICK) {
+            final Vector2 worldSpaceCoordinates = projector.unproject(x, y);
+            final Actor gameContainer = toolbarWidget.getParent();
+            final float offSetX = gameContainer.getX();
+            final float offSetY = gameContainer.getY() + toolbarWidget.getHeight();
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java:52]: style

It's offset not offSet

File: src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    /**
+     * Set the game widget used to get the offset.
+     *
+     * @param toolbarWidget The widget in which the game is rendered.
+     */
+    public void setToolbarWidget(final Actor toolbarWidget) {
+        this.toolbarWidget = toolbarWidget;
+    }
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java:67]: Why not just set the offset?

File: src/main/java/com/thebois/controllers/toolbar/ToolbarController.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+        toolbarView = createToolbarView();
+
+        placeStructureController.setToolbarWidget(toolbarView.getPane());
+    }
+
+    private ToolbarView createToolbarView() {
+        final List<IActorView> views = controllers.stream().map(IController::getView).collect(
+            Collectors.toList());
+        return new ToolbarView(views);
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/controllers/toolbar/ToolbarController.java:48]: Clean 👍

File: src/main/java/com/thebois/controllers/toolbar/ToolbarController.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    public ToolbarView getView() {
+        return toolbarView;
+    }
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/controllers/toolbar/ToolbarController.java:54]: Wasn't one of the points of having the controllers create the views that the outside world doesn't need to know about the view? They just need the `IView`, right?

File: src/main/java/com/thebois/listeners/events/OnClickEvent.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package com.thebois.listeners.events;
+
+/**
+ * Sends out information about a value when notifier has been clicked.
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/listeners/events/OnClickEvent.java:4]: >when notifier has been clicked

What's a notifier? I suppose you're avoiding using "button" since it is more generic than that and can be used for things other than buttons. I'd rephrase the entire javadoc to not describe it as if it is sending out any information, because this is actually just a data-carrier object, it doesn't send anything.

"Contains information about a value that has been clicked." or something similar, better describes it imo.

File: src/main/java/com/thebois/views/GameScreen.java [hunk 9]
```diff
@@ -22,31 +23,34 @@
     private final OrthographicCamera camera;
     private final FitViewport viewport;
     private final Color backgroundColor = new Color(0, 0, 0.2f, 1);
-    private final Skin skin;
-    private final GameView gameView;
+    private final Actor gameView;
     private final InfoView infoView;
+    private final ToolbarView toolbarView;
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/GameScreen.java:28]: If you're making `gameView` an `Actor` then I think all three of them should be `Actor`s.

File: src/main/java/com/thebois/views/GameScreen.java [hunk 10]
```diff
@@ -55,14 +59,20 @@ private void createStage() {
...
 
-        final Table infoAndWorldGroup = new Table();
-        infoAndWorldGroup.setFillParent(true);
-        stage.addActor(infoAndWorldGroup);
-
-        infoAndWorldGroup.left();
+        final Table infoAndGameTable = new Table();
+        infoAndGameTable.setFillParent(true);
+        infoAndGameTable.add(infoView.getPane()).fillY().expandY();
+        infoAndGameTable.add(gameContainer);
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/GameScreen.java:65]: Having `gameContainer` be something that contains both the `gameView` (which is actually closer to `worldView` now, right?) and the `toolbarView` makes it sound like the toolbar is a part of the game. 

Since the infoview, gameview and toolbarview are all three elements on the same abstraction level (they all contain sub-views and represent a section of the game window), they should all be put directly into the table. 

Tables are like grids, you should be able to set up two rows and two columns and have the infoview take up two rows, while the others take up one. And then it should be renamed to something more appropriate like `userInterfaceTable` or something.

File: src/main/java/com/thebois/views/game/StructureButton.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * Instantiate a structure button with the type of structure it should represent.
+     *
+     * @param structureType The type or structure to represent.
+     * @param skin          The skin used to display the button.
+     */
+    public StructureButton(final StructureType structureType, final Skin skin) {
+        super(skin, "toggle");
+        button = this;
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/game/StructureButton.java:39]: What are you doing here? I don't understand.

You can just use `this` directly, you don't need to store it in a field.

File: src/main/java/com/thebois/views/game/StructureButton.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+                }
+            }
+        });
+    }
+
+    private void notifyListeners(final StructureType type) {
+        final OnClickEvent<StructureType> event = new OnClickEvent<>(type);
+        for (final IEventListener<OnClickEvent<StructureType>> listener : onClickListeners) {
+            listener.onEvent(event);
+        }
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/game/StructureButton.java:71]: style

Can be simplified to
```java
onClickListeners.forEach(listener -> listener.onEvent(event));
```

File: src/main/java/com/thebois/views/toolbar/StructureToolbarView.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * View that renders structure buttons used for selecting what structure to build.
+ */
+public class StructureToolbarView implements IActorView {
+
+    private final Actor structureButtons;
+
+    /**
+     * Instansiate a table with the different structure buttons.
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/toolbar/StructureToolbarView.java:22]: This isn't a table. Just "Instantiates with buttons for each structure."

typo: Instansiate -> Instantiate

File: src/main/java/com/thebois/views/toolbar/StructureToolbarView.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        structureButtons = createToolbar(skin, listener);
+    }
+
+    private Actor createToolbar(
+        final Skin skin, final IEventListener<OnClickEvent<StructureType>> listener) {
+
+        // Set up table.
+        final Table toolbarTable = new Table();
+        toolbarTable.left().top();
+        toolbarTable.row().expandY().fillY();
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/toolbar/StructureToolbarView.java:38]:
```java
final Table toolbarTable = createTable();
```

File: src/main/java/com/thebois/views/toolbar/StructureToolbarView.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+        // Button group settings.
+        final ButtonGroup<StructureButton> buttonGroup = new ButtonGroup<>();
+        buttonGroup.setMaxCheckCount(1);
+        buttonGroup.setMinCheckCount(1);
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/toolbar/StructureToolbarView.java:43]:
```java
final ButtonGroup<StructureButton> buttonGroup = createButtonGroup();
```

File: src/main/java/com/thebois/views/toolbar/StructureToolbarView.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+        // Add buttons to table and button group.
+        for (final StructureType type : StructureType.values()) {
+            final StructureButton structureButton = new StructureButton(type, skin);
+            toolbarTable.add(structureButton);
+            buttonGroup.add(structureButton);
+            structureButton.registerListener(listener);
+        }
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/toolbar/StructureToolbarView.java:51]:
```java
createStructureButtons();
```

File: src/main/java/com/thebois/views/toolbar/StructureToolbarView.java [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+import com.thebois.models.world.structures.StructureType;
+import com.thebois.views.game.StructureButton;
+import com.thebois.views.info.IActorView;
+
+/**
+ * View that renders structure buttons used for selecting what structure to build.
+ */
+public class StructureToolbarView implements IActorView {
+
+    private final Actor structureButtons;
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/toolbar/StructureToolbarView.java:19]: style

This isn't the buttons themselves, this is a table containing the buttons. So it's a sort of root-element, a container of structure buttons. So maybe `structureButtonContainer`.

File: src/main/java/com/thebois/views/toolbar/StructureToolbarView.java [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+import com.thebois.listeners.IEventListener;
+import com.thebois.listeners.events.OnClickEvent;
+import com.thebois.models.world.structures.StructureType;
+import com.thebois.views.game.StructureButton;
+import com.thebois.views.info.IActorView;
+
+/**
+ * View that renders structure buttons used for selecting what structure to build.
+ */
+public class StructureToolbarView implements IActorView {
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/toolbar/StructureToolbarView.java:17]: style

This is more of a `BuildMenuToolbarView`, right? It's not a view of structures, it's a view of a menu that allows for selecting of which structure to build.

File: src/main/java/com/thebois/views/toolbar/StructureToolbarView.java (hunk 14 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/toolbar/StructureToolbarView.java:38]: If it's unclear, I mean that you need to do some functional decomposition here.

File: src/main/java/com/thebois/views/toolbar/ToolbarView.java [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * The view that displays the different items in the toolbar.
+ */
+public class ToolbarView {
+
+    private final HorizontalGroup toolbar;
+
+    /**
+     * Instantiates a new view of a toolbar that the player can interact with.
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/toolbar/ToolbarView.java:16]: style

Nothing about this view allows for interaction, it's the different sub-views that do that.

File: src/main/java/com/thebois/views/toolbar/ToolbarView.java [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+    public ToolbarView(final Iterable<IActorView> widgetViews) {
+
+        toolbar = new HorizontalGroup();
+        toolbar.expand().fill().expand();
+        for (final IActorView actorView : widgetViews) {
+            toolbar.addActor(actorView.getWidgetContainer());
+        }
+    }
+
+    public Actor getPane() {
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/toolbar/ToolbarView.java:29]: Missing `@Override`

**Review by martinjonsson01 — APPROVED** [review]
Looking good. 

Is it just me or can the buttons be quite a lot smaller? It's not a touch game, so the buttons should be able to be quite small (allowing for a larger view of the game world).

File: src/main/java/com/thebois/views/GameScreen.java (hunk 9 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/GameScreen.java:28]: At the moment the other 2 are not actors while `GameView` has always been an actor. I guess that the other 2 could be made into `IActorViews` if we want that. Or I change `gameView` back to `GameView`

File: src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java (hunk 5 above)
Comment (Matte02) [reply src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java:67]: When this method is being called, the widget itself doesn't have a height or width. I couldn't figure out why it didn't, but I think it's because it has not been rendered yet which it does in the first render loop.

File: src/main/java/com/thebois/views/GameScreen.java (hunk 9 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/GameScreen.java:28]: Ah, they should all ideally be extending `Actor`, but that's another refactor so let's just leave it as it is right now.

File: src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java (hunk 5 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java:67]: Oh, well that's unfortunate. Can't you inject the toolbar in the constructor instead, then?

File: src/main/java/com/thebois/views/game/StructureButton.java (hunk 11 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/game/StructureButton.java:39]: Fair enough I just thought `button` made it slightly more readable, but maybe `this` is better, I can change it back to this.

File: src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java (hunk 5 above)
Comment (Matte02) [reply src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java:67]: Nope because the toolbar is created after the controllers.

File: src/main/java/com/thebois/views/game/StructureButton.java (hunk 11 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/game/StructureButton.java:39]: Ah, I see. The reason this confuses me a lot is because it looks like you're referring to some other button-object, when in fact you're referring to `this` button instance. Since the class *is* a button, saying `this.isToggled` should be easy enough to understand.

File: src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java (hunk 5 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/controllers/toolbar/PlaceStructureController.java:67]: Welp, guess it'll have to stay like this then. Still feels a bit wrong, but at least it works!

File: src/main/java/com/thebois/Pawntastic.java (hunk 1 above)
Comment (Matte02) [reply src/main/java/com/thebois/Pawntastic.java]: Apparently, it was not needed anywhere only for calculating the tile size and indirectly the game view size. 
And then just have everything just expand and fill.

File: src/main/java/com/thebois/controllers/toolbar/ToolbarController.java (hunk 7 above)
Comment (Matte02) [reply src/main/java/com/thebois/controllers/toolbar/ToolbarController.java:54]: Gonna quote you 
> Ah, they should all ideally be extending Actor, but that's another refactor so let's just leave it as it is right now.

I could easily make them(ToolbarView and InfoView) in IActorViews, or just leave them like this. Because making them into actors would take som time.

File: src/main/java/com/thebois/views/game/StructureButton.java (hunk 11 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/game/StructureButton.java]: I now remembered one more reason why I did the change use `button`, `this` in the lambda function further down doesn't work.
But I've made some changes, tell me what you think.

File: src/main/java/com/thebois/controllers/game/StructureController.java (hunk 3 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/controllers/game/StructureController.java:22]: Yeah, I've noticed it myself a few times, it's quite annoying really.

File: src/main/java/com/thebois/views/toolbar/ToolbarView.java (hunk 20 above)
Comment (Matte02) [reply src/main/java/com/thebois/views/toolbar/ToolbarView.java:34]: ???
Toolbar view doesn't implement anything, just like info view.

Comment (martinjonsson01) [reply src/main/java/com/thebois/views/toolbar/ToolbarView.java:34]: They both really should... Guess we'll have to write that in the SDD, then.

File: src/main/java/com/thebois/controllers/toolbar/ToolbarController.java (hunk 7 above)
Comment (Matte02) [reply src/main/java/com/thebois/controllers/toolbar/ToolbarController.java:54]: Not all of them, but I could make InfoView and ToolbarView into IActorViews

Comment (martinjonsson01) [reply src/main/java/com/thebois/controllers/toolbar/ToolbarController.java:54]: Why not `GameView`?

If it can't be done uniformly, then let's not bother.

Comment (martinjonsson01) [reply src/main/java/com/thebois/controllers/toolbar/ToolbarController.java]: >then let's not bother

Oh, well if it was that easy, then just leave it like you've done now.

File: src/main/java/com/thebois/views/GameScreen.java [hunk 21]
```diff
@@ -40,7 +40,7 @@
     public GameScreen(
         final FitViewport viewport,
         final OrthographicCamera camera,
-        final Actor gameView,
+        final IActorView gameView,
         final IActorView infoView,
         final IActorView toolbarView) {
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/views/GameScreen.java]: Looking much better now 👍

**Review by martinjonsson01 — APPROVED** [review]
Nicely done!

**Review by jonathan-lindqvist — APPROVED** [review]
LGTM
